content/doc/book/pipeline/pipeline-best-practices.adoc

```diff
@@ -0,0 +1,53 @@
+... Solution: Instead of using JsonSlurper, use a shell step and return the standard out. This shell would look something like this: `def JsonReturn = sh label: '', returnStdout: true, script: 'echo "$LOCAL_FILE"| jq "."'`. This will use agent resources to read the file and even parse the file with a good `jq` filter
+.. HttpRequest: Many times this would be used to grab data from an external source and store it into a variable. This is not a good idea to use because not only is that request coming from the master(which could give incorrect results for things like HTTPS requests if the master does not have certificates loaded) but also the return is again stored twice.
+... Solution: Again using the shell step but this time with a `curl` command is the best direction to go.Then have the request come from master and also the data size can be shrunk with filtering.
+. **Pipeline functions as Glue:** Make sure to use Pipeline steps as glue rather than Main functionality. This means that rather than relying on pipeline functionality(groovy or pipeline steps) to drive the build process forward, it is better to use single steps(such as `sh`) to accomplish multiple parts of the build. The less pipeline functions used, the less memory will consumed up on master
+.. Example: Using a single maven build step to drive the build through its build/test/deploy process.
+. **Just say no to Custom Pipeline Steps:** Wherever possible stay away from Customized/overwritten Pipeline Steps. This is the process where the standard pipeline APIs like `sh` or `timeout` are overwritten using shared libraries
+.. This is dangerous because the pipeline APIs can change at any time causing custom code to break or give different results than expected
+.. It can become difficult to troubleshoot when something is breaking or causing issues
+.. Because of the ubiquitous use of these steps throughout pipelines, if something is coded incorrectly/inefficiently the results can be catastrophic to Jenkins
+. **Reduce Repetition of Pipeline Steps:** Combine pipeline steps into single steps as often as possible to reduce the size of the flow-node graph. 
+.. For each pipeline step that is executed there is a flow-node graph that allows a pipeline to know what step it is currently executing and when resuming, the flow node graph has to be loaded onto the stack
+.. This means whenever doing a series of `echo` or `sh` steps make sure to combine them into a single step or script rather than individual steps
+. **Dealing with Concurrency within pipelines:** Try not to share workspaces across multiple pipeline executions or multiple distinct pipelines. 
```

Comment:
`Try not to share workspaces across multiple pipeline executions`, whats the alternative if parallels branches want to access worksapce? stash/unstash?

Reply:
I should add something about this but ideally yes. Only stash/unstash things you need not the whole workspace